src/main/kotlin/org/rust/lang/core/resolve2/HangingModData.kt

```diff
@@ -0,0 +1,158 @@
+        isEnabledByCfgInner = scope.isEnabledByCfg(crate),
+        fileId = null,
+        fileRelativePath = "",
+        ownedDirectoryId = null,
+        hasPathAttribute = false,
+        hasMacroUse = false,
+        isNormalCrate = false,
+        context = contextData,
+        crateDescription = "block in ${contextData.crateDescription}",
+    )
+
+    val collectorContext = CollectorContext(crate, project, isHangingMode = true)
+    val modCollectorContext = ModCollectorContext(defMap, collectorContext)
+    val dollarCrateHelper = createDollarCrateHelper(scope)
+    collectScope(scope, hangingModData, modCollectorContext, dollarCrateHelper = dollarCrateHelper)
+
+    val indicator = ProgressManager.getGlobalProgressIndicator() ?: EmptyProgressIndicator()
+    DefCollector(project, defMap, collectorContext, pool = null, indicator).collect()
+    return hangingModData
+}
+
+private val HANGING_MOD_DATA_KEY: Key<Pair<ModData, List<Long>>> = Key.create("HANGING_MOD_DATA_KEY")
+
+private fun RsElement.getContextModInfo(): RsModInfoBase {
+    val context = contextStrict<RsItemsOwner>() ?: return InfoNotFound
+    return when {
+        context is RsMod -> getModInfo(context)
+        context is RsBlock && shouldCreateHangingModInfo(context) -> getHangingModInfo(context)
+        else -> context.getContextModInfo()
+    }
+}
+
+private fun shouldCreateHangingModInfo(scope: RsItemsOwner): Boolean =
+    scope.itemsAndMacros.any { it is RsItemElement || it is RsMacro || it is RsMacroCall }
+
+private fun createDollarCrateHelper(scope: RsItemsOwner): DollarCrateHelper? {
+    val call = scope.findMacroCallExpandedFromNonRecursive() as? RsMacroCall ?: return null
+    val expansion = call.expansion ?: return null
+    val expansionText = expansion.file.text
+    if (!expansionText.contains(MACRO_DOLLAR_CRATE_IDENTIFIER)) return null
+
+    val defCrate = call.resolveToMacroAndGetContainingCrate()?.id ?: return null
+    val rangesInExpansion = MACRO_DOLLAR_CRATE_IDENTIFIER_REGEX.findAll(expansionText)
+        .associate { it.range.first to defCrate }
+    // TODO: Proper implementation with support of `defHasLocalInnerMacros`, macro expanded to macro, etc
+    return DollarCrateHelper(expansion.ranges, rangesInExpansion, defHasLocalInnerMacros = false, defCrate)
+}
+
+private class LocalScopeDataPsiHelper(
+    private val scope: RsItemsOwner,
+    private val modData: ModData,
+    private val delegate: DataPsiHelper?,
+) : DataPsiHelper {
+    override fun psiToData(scope: RsItemsOwner): ModData? {
+        if (scope == this.scope) return modData
+        return delegate?.psiToData(scope)
+    }
+
+    override fun dataToPsi(data: ModData): RsItemsOwner? {
+        if (data == modData) return scope
+        data.path.getRelativePathTo(modData.path)?.let { relativePath ->
+            return relativePath.fold(scope as RsItemsOwner?) { scope, segment ->
+                scope?.getChildModule(segment)
+            }
```

Comment:
It will be needed for local modules (which is not currently working because `resolvePathFp` doesn't support local modules). I added test for resolve of local module, where this code is used.